ethcore/builtin/src/lib.rs

```diff
@@ -207,6 +218,93 @@ impl ModexpPricer {
 	}
 }
 
+/// Bls12 pairing price
+#[derive(Debug, Copy, Clone)]
+struct Bls12PairingPrice {
+	base: u64,
+	pair: u64,
+}
+
+/// bls12_pairing pricing model. This computes a price using a base cost and a cost per pair.
+#[derive(Debug)]
+struct Bls12PairingPricer {
+	price: Bls12PairingPrice,
+}
+
+/// Pricing for constant Bls12 operations (ADD and MUL in G1 and G2, as well as mappings)
+#[derive(Debug, Copy, Clone)]
+pub struct Bls12ConstOperationsPricer {
+	/// Fixed price.
+	pub price: u64,
+}
+
+/// Discount table for multiexponentiation (Peppinger algorithm)
+/// Later on is normalized using the divisor
+pub const BLS12_MULTIEXP_DISCOUNTS_TABLE: [[u64; 2]; BLS12_MULTIEXP_PAIRS_FOR_MAX_DISCOUNT] = [
+	[1, 1200], [2, 888], [3, 764], [4, 641], [5, 594], [6, 547], [7, 500], [8, 453], 
+	[9, 438], [10, 423], [11, 408], [12, 394], [13, 379], [14, 364], [15, 349], [16, 334], 
+	[17, 330], [18, 326], [19, 322], [20, 318], [21, 314], [22, 310], [23, 306], [24, 302], 
+	[25, 298], [26, 294], [27, 289], [28, 285], [29, 281], [30, 277], [31, 273], [32, 269], 
+	[33, 268], [34, 266], [35, 265], [36, 263], [37, 262], [38, 260], [39, 259], [40, 257], 
+	[41, 256], [42, 254], [43, 253], [44, 251], [45, 250], [46, 248], [47, 247], [48, 245], 
+	[49, 244], [50, 242], [51, 241], [52, 239], [53, 238], [54, 236], [55, 235], [56, 233], 
+	[57, 232], [58, 231], [59, 229], [60, 228], [61, 226], [62, 225], [63, 223], [64, 222], 
+	[65, 221], [66, 220], [67, 219], [68, 219], [69, 218], [70, 217], [71, 216], [72, 216], 
+	[73, 215], [74, 214], [75, 213], [76, 213], [77, 212], [78, 211], [79, 211], [80, 210], 
+	[81, 209], [82, 208], [83, 208], [84, 207], [85, 206], [86, 205], [87, 205], [88, 204], 
+	[89, 203], [90, 202], [91, 202], [92, 201], [93, 200], [94, 199], [95, 199], [96, 198], 
+	[97, 197], [98, 196], [99, 196], [100, 195], [101, 194], [102, 193], [103, 193], [104, 192], 
+	[105, 191], [106, 191], [107, 190], [108, 189], [109, 188], [110, 188], [111, 187], [112, 186], 
+	[113, 185], [114, 185], [115, 184], [116, 183], [117, 182], [118, 182], [119, 181], [120, 180], 
+	[121, 179], [122, 179], [123, 178], [124, 177], [125, 176], [126, 176], [127, 175], [128, 174]
+];
+
+/// Max discount allowed
+pub const BLS12_MULTIEXP_MAX_DISCOUNT: u64 = 174;
+/// Max discount is reached at this number of pairs
+pub const BLS12_MULTIEXP_PAIRS_FOR_MAX_DISCOUNT: usize = 128;
+/// Divisor for discounts table
+pub const BLS12_MULTIEXP_DISCOUNT_DIVISOR: u64 = 1000;
+/// Length of single G1 + G2 points pair for pairing operation
+pub const BLS12_G1_AND_G2_PAIR_LEN: usize = SERIALIZED_G1_POINT_BYTE_LENGTH + SERIALIZED_G2_POINT_BYTE_LENGTH;
+
+/// Pricing for constant Bls12 operations (ADD and MUL in G1 and G2, as well as mappings)
+#[derive(Debug, Copy, Clone)]
+pub struct Bls12MultiexpPricer {
+	/// Base const of the operation (G1 or G2 multiplication)
+	pub base_price: Bls12ConstOperationsPricer,
+	/// Length of input per one pair (point + scalar)
+	pub len_per_pair: usize,
```

Comment:
@shamatar, as I see, `Bls12MultiexpPricer.len_per_pair` should be a constant in the chain specs. I think that we should avoid this constant only useful to reuse internal structures. What about split it into  `Bls12G1MultiexpPricer` and  `Bls12G2MultiexpPricer` ?

Reply:
Right now part of the chainspec looks like
```
"price": { "bls12_multiexp_price": { "base_price": 12000, "len_per_pair": 96 }}
```

Would it be better to instead change `len_per_pair` to some constant like `G2`? If you want to avoid any implementation or logical configuration options in the chainspec than I would rather try make a pricer templated over some market trait that indicates G1/G2 and see if it would work. Worst case two separate structures will be a final resort.